docs/code-generation.md

```diff
@@ -0,0 +1,135 @@
+In fact, the CRDs for individual AWS services like S3 or RDS were thin wrappers
+around CloudFormation stacks that described the object being operated upon.
+
+[1]: https://github.com/amazon-archives/aws-service-operator/tree/master/code-generation
+[2]: https://github.com/amazon-archives/aws-service-operator/tree/master/models
+[3]: https://github.com/amazon-archives/aws-service-operator/tree/master/code-generation/pkg/codegen/assets
+[4]: https://github.com/amazon-archives/aws-service-operator/blob/b4befd62322a57ac78aa39ea08771fc32912592a/code-generation/pkg/codegen/assets/aws-service-operator.yaml.templ#L13-L31
+[5]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/operator.go.templ
+[6]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/types.go.templ
+
+kudo is a platform for building Kubernetes Operators. It stores state in its
+own kudo.dev CRDs and allows users to define "plans" for a deployed application
+to deploy itself. We determined that kudo was not a particularly good fit for
+ASO for a couple reasons. First, we needed a way to generate CRDs in several
+API groups (s3.aws.com and iam.aws.com for example) and the ASO controller code
+isn't deploying an "application" that needs to have a controlled deployment
+plan. Instead, ASO is a controller that facilitates creation and management of
+various AWS service objects using Kubernetes CRD instances.
+
+kubebuilder is the the recommended upstream tool for generating CRDs and
+controller stub code. It is a Golang binary that creates the scaffolding for
+CRDs and controller Golang code. Unfortunately, it doesn't really handle
+multiple API groups in the same source code repository, and we need that for
+ASO. For example, we will have an s3.aws.com and an iam.aws.com API group with
+CRDs representing those different API objects and interfaces.
+
+That said, much of kubebuilder simply sets up some templates and then runs the
+`controller-gen` binary from the sigs.kubernetes.io/controller-tools repository
+to generate Golang code.
+
+Our final option was to build a hybrid custom code generator that used
+controller-runtime under the hood but allowed us to generate controller stub
+code for multiple API groups and place generated code [in directories][7] that
+represented Golang best practices. This option gives us the flexibility to
+generate the files and content for multiple API groups but still stay within
+the recommended guardrails of the upstream Kubernetes community.
+
+[7]: https://github.com/kubernetes-sigs/kubebuilder/issues/1268
+
+## Hybrid custom+controller-runtime proposal
+
+![Multi-phase approach to code generation for ASO](images/multi-phase-code-generation.png)
+
+This approach uses multiple phases of code generation.
+
+The first code generation phase consumes model information from a canonical
```

Comment:
I'm not entirely sure, hence a clarification question: are you talking about using Go pragma, esp. the `//go:generate` directive for [generating](https://blog.golang.org/generate) this?

Reply:
Yes, the first phase of code generation can be a Go file that reads the model files from `aws-sdk-go/models` and creates the Go type files that are used as input for the second phase of code generation.